pisa/stages/osc/layers.py

```diff
@@ -47,170 +46,141 @@ def __call__(self, *args):
-    shape = (np.int64(len(cz)), np.int64(max_layers))
-    n_layers = np.zeros(shape[0], dtype=np.int32)
-    distance = np.zeros(shape=shape, dtype=FTYPE)
-    density = np.zeros(shape=shape, dtype=FTYPE)
+
+    # The densities, distances and number of layers are appended to one long list
+    # which is later reshaped into containers of size (# of cz values, max_layers)
+    # in the pi_prob3 module
+
+    densities = np.zeros((len(cz), max_layers), dtype=FTYPE)
+    distances = np.zeros((len(cz), max_layers), dtype=FTYPE)
+    number_of_layers = np.zeros(len(cz))
 
     # Loop over all CZ values
-    for k, coszen in enumerate(cz):
-        tot_earth_len = -2 * coszen * r_detector
-
-        # To store results
-        traverse_rhos = np.zeros(max_layers, dtype=FTYPE)
-        traverse_dist = np.zeros(max_layers, dtype=FTYPE)
-        traverse_electron_frac = np.zeros(max_layers, dtype=FTYPE)
-
-        # Above horizon
-        if coszen >= 0:
-            kappa = (detector_depth + prop_height)/r_detector
-            path_len = (
-                r_detector * np.sqrt(coszen**2 - 1 + (1 + kappa)**2)
-                - r_detector * coszen
-            )
-
-            # Path through the air:
-            kappa = detector_depth / r_detector
-            lam = (
-                coszen + np.sqrt(coszen**2 - 1 + (1 + kappa) * (1 + kappa))
-            )
-            lam *= r_detector
-            path_thru_atm = (
-                prop_height * (prop_height + 2*detector_depth + 2*r_detector)
-                / (path_len + lam)
-            )
-            path_thru_outerlayer = path_len - path_thru_atm
-            traverse_rhos[0] = 0.0
-            traverse_dist[0] = path_thru_atm
-            traverse_electron_frac[0] = default_elec_frac
-
-            # In that case the neutrino passes through some earth (?)
-            layers = 1
-            if detector_depth > min_detector_depth:
-                traverse_rhos[1] = rhos[0]
-                traverse_dist[1] = path_thru_outerlayer
-                traverse_electron_frac[1] = YeFrac[-1]
-                layers += 1
-
-        # Below horizon
+    for i, coszen in enumerate(cz):
+
+        r_prop = r_detector+detector_depth+prop_height
+        # Compute the full path length
+        # path_len = -r_detector * coszen + np.sqrt(r_detector**2. * coszen**2 - (r_detector**2. - r_prop**2.))
+
+        # Determine if there will be a crossing of layer
+        # idx is the index of the first inner layer
+        idx = np.where(radii<r_detector)[0][0]
+
+        # Deal with paths that do not have tangeants
+        if coszen >= coszen_limit[idx]: 
+            cumulative_distances = -r_detector * coszen + np.sqrt(r_detector**2. * coszen**2. - r_detector**2. + radii[:idx]**2.)
+            # a bit of flippy business is done here to order terms
+            # such that numpy diff can work
+            segments_lengths = np.diff(np.concatenate((np.array([0.], dtype=FTYPE), cumulative_distances[::-1])))
```

Comment:
It is a bit unfortunate that all of this is done with several concatenations, because that is a very slow operation where the entire array has to be copied. Could you rewrite this with for-loops? I think that it would make the whole thing also easier to understand because you wouldn't have to do all these inversions and concatenations, and will probably solve your ftype problems as well.

Reply:
For now I believe that the most urgent issue is to get a PR that works merged, so that Philipp can move PISA4 onward. Any issues regarding performances can be addressed by a new PR in the new version of PISA.

Reply:
Agreed!

Reply:
Okay, I made a PR to this PR, @bourdeet ! It fixes the FTYPE issue and all the unit tests pass, so this should be all good now I hope. :)